services/ui_backend_service/api/heartbeat_monitor.py

```diff
@@ -63,25 +99,88 @@ async def get_run(self, run_id):
                           )
     return result.body if result.response_code==200 else None
   
-  async def load_and_broadcast(self, run_id):
-    '''
-    Load updated run and broadcast the changes to the websocket event handler.
-    '''
-    run = await self.get_run(run_id)
+  async def load_and_broadcast(self, key):
+    run = await self.get_run(key)
     resources = resource_list(self._run_table.table_name, run)
     self.event_emitter.emit('notify', 'UPDATE', resources, run)
 
-  async def check_heartbeats(self):
-    '''
-    Async Task that is responsible for checking the heartbeats of all monitored runs,
-    and triggering handlers in case the heartbeat is too old.
-    '''
-    while True:
-      time_now = int(datetime.datetime.utcnow().timestamp()) # same format as the metadata heartbeat uses
-      # print(f"checking heartbeats of {len(self.watched)} runs, timestamp: {time_now}", flush=True)
-      for run_number, last_heartbeat_ts in list(self.watched.items()):
-        if time_now - last_heartbeat_ts > HEARTBEAT_INTERVAL * 2:
-          await self.load_and_broadcast(run_number)
-          self.remove_run_from_watch(run_number)
-          
-      await asyncio.sleep(HEARTBEAT_INTERVAL)
+class TaskHeartbeatMonitor(HeartbeatMonitor):
+  '''
+  Service class for adding objects with heartbeat timestamps to be monitored and acted upon
+  when heartbeat becomes too old.
+
+  Starts an async watcher loop that periodically checks the heartbeat of the subscribed tasks for expired timestamps
+
+  Usage
+  -----
+  Responds to event_emitter emissions with messages:
+    "task-heartbeat", "update", data -> updates heartbeat timestamp that is found in database
+    "task-heartbeat", "complete" data -> removes task from heartbeat checks
+  '''
+  def __init__(self, event_emitter=None):
+    # Init the abstract class
+    super().__init__(
+      event_name="task-heartbeat",
+      event_emitter=event_emitter
+    )
+    # Table for data fetching for load_and_broadcast and add_to_watch
+    self._task_table = AsyncPostgresDB.get_instance().task_table_postgres
+
+  async def heartbeat_handler(self, action, data):
+    if action == "update":
+      await self.add_to_watch(data)
+    elif action == "complete":
+      key = self.generate_dict_key(data)
+      self.remove_from_watch(key)
+    
+  async def add_to_watch(self, data):
+    # TODO: Optimize db trigger so we do not have to fetch a record in order to add it to the
+    # heartbeat monitor
+    task = await self.get_task(
+      data["flow_id"],
+      data["run_number"],
+      data["step_name"],
+      data["task_id"]
+    )
```

Comment:
There is a performance concern related to the DB trigger_notify not carrying all necessary info in order to add a task to be monitored. Critically the `attempt_id` is missing from the trigger, so we need to perform a db fetch because of this.